x/concentrated-liquidity/incentives.go

```diff
@@ -270,3 +281,62 @@ func (k Keeper) GetIncentiveRecord(ctx sdk.Context, poolId uint64, denom string,
 func (k Keeper) GetAllIncentiveRecordsForPool(ctx sdk.Context, poolId uint64) ([]types.IncentiveRecord, error) {
 	return osmoutils.GatherValuesFromStorePrefixWithKeyParser(ctx.KVStore(k.storeKey), types.KeyPoolIncentiveRecords(poolId), ParseFullIncentiveRecordFromBz)
 }
+
+// GetUptimeGrowthInsideRange returns the uptime growth within the given tick range for all supported uptimes
+func (k Keeper) GetUptimeGrowthInsideRange(ctx sdk.Context, poolId uint64, lowerTick int64, upperTick int64) ([]sdk.DecCoins, error) {
+	pool, err := k.getPoolById(ctx, poolId)
+	if err != nil {
+		return []sdk.DecCoins{}, err
+	}
+
+	// Get global uptime accumulator values
+	globalUptimeValues, err := k.getUptimeAccumulatorValues(ctx, poolId)
+	if err != nil {
+		return []sdk.DecCoins{}, err
+	}
+
+	// Get current, lower, and upper ticks
+	currentTick := pool.GetCurrentTick().Int64()
+	lowerTickInfo, err := k.getTickInfo(ctx, poolId, lowerTick)
+	if err != nil {
+		return []sdk.DecCoins{}, err
+	}
+	upperTickInfo, err := k.getTickInfo(ctx, poolId, upperTick)
+	if err != nil {
+		return []sdk.DecCoins{}, err
+	}
+
+	// Calculate uptime growth between lower and upper ticks
+	// Note that we regard "within range" to mean [lowerTick, upperTick),
+	// inclusive of lowerTick and exclusive of upperTick.
+	if currentTick < lowerTick {
+		// If current tick is below range, we subtract uptime growth of upper tick from that of lower tick
+		return osmoutils.SubDecCoinArrays(getUptimeTrackerValues(lowerTickInfo.UptimeTrackers), getUptimeTrackerValues(upperTickInfo.UptimeTrackers))
+	} else if currentTick < upperTick {
+		// If current tick is within range, we subtract uptime growth of lower and upper tick from global growth
+		globalMinusUpper, err := osmoutils.SubDecCoinArrays(globalUptimeValues, getUptimeTrackerValues(upperTickInfo.UptimeTrackers))
+		if err != nil {
+			return []sdk.DecCoins{}, err
+		}
+
+		return osmoutils.SubDecCoinArrays(globalMinusUpper, getUptimeTrackerValues(lowerTickInfo.UptimeTrackers))
+	} else {
+		// If current tick is above range, we subtract uptime growth of lower tick from that of upper tick
+		return osmoutils.SubDecCoinArrays(getUptimeTrackerValues(upperTickInfo.UptimeTrackers), getUptimeTrackerValues(lowerTickInfo.UptimeTrackers))
+	}
```

Comment:
This calculation model looks quite different from fees where we do the following:
1. calculate fee growth above upper tick
2. calculate fee growth below lower tick

This gets us fee growth outside.

Subtract fee growth outside from global fee growth to get fee growth inside.

From my understanding, we should be able to compute the uptime growth using a similar model.

Is it possible to use a similar strategy to reduce the complexity stemming from various strategies when a similar one could be used?

Reply:
With fees, we also get fee growth outside first to then compute fee growth inside by subtracting it from global. With incentives, it's the opposite which introduces additional complexity from having to ponder why this is the case.

Reply:
To be clear, I don't mind having this approach for computing growth inside and outside.

What I'm advocating for is consistency between fee growth and incentives uptime tracker growth

Reply:
Discussed on call that this was not blocking but worth coming back to: https://github.com/osmosis-labs/osmosis/issues/4453